lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -311,3 +312,83 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
   }
   assert(isValid());
 }
+
+/// Derive a concrete type of self and conformance from protocol conformance
+/// analysis If successful, initializes a valid ConcreteExistentialInfo.
+ConcreteExistentialInfo::ConcreteExistentialInfo(
+    ProtocolConformanceAnalysis *PCA, ClassHierarchyAnalysis *CHA,
+    SILBuilder &Builder, ProtocolDecl *Protocol, FullApplySite Apply,
+    SILValue &Arg, Operand &ArgOperand) {
+
+  // Determine the sole conforming type.
+  auto *NTD = PCA->findSoleConformingType(Protocol);
+  if (!NTD)
+    return;
+
+  // Sole conforming class should not be public or have derived classes.
+  ClassDecl *CD;
+  if ((CD = dyn_cast<ClassDecl>(NTD)) &&
+      (CD->getEffectiveAccess() == AccessLevel::Open ||
+       CHA->hasKnownDirectSubclasses(CD) ||
+       CHA->hasKnownIndirectSubclasses(CD))) {
+    return;
+  }
+
+  // findInitExistential for the argument.
+  isCopied = false;
+  OpenedArchetypeDef = Arg;
+  InitExistential = findInitExistential(ArgOperand, OpenedArchetype,
+                                        OpenedArchetypeDef, isCopied);
+
+  // If no InitExistential is found, check if the argument is already an opened
```

Comment:
Sorry, I'm really confused now. The code below runs regardless of whether `InitExistential` was found or not. How is it related to `InitExstential`?

Reply:
Dumb question: if the caller already knows the ConcreteType and Protocol, why do we need to bother looking for the `InitExistential`?

Reply:
I still need the OpenedArchetypeDef to be able to cast it to the concrete type. findInitExistential helps me determine this OpenedArchetypeDef. For cases where OpenedArchetypeDef can not be found (because there was no InitExistential), then the code below does extra effort in trying to find if Arg is itself OpenedArcheTypeDef or AllocStackInst, in which case it determines the OpenedArcheTypeDef. I will update the comment. Let me know if this makes sense.